include/grpc++/impl/codegen/call.h

```diff
@@ -523,14 +525,30 @@ class CallOpClientRecvStatus {
 
 /// An abstract collection of CallOpSet's, to be used whenever
 /// CallOpSet objects must be thought of as a group. Each member
-/// of the group should have a shared_ptr back to the collection,
-/// as will the object that instantiates the collection, allowing
-/// for automatic ref-counting. In practice, any actual use should
-/// derive from this base class. This is specifically necessary if
-/// some of the CallOpSet's in the collection are "Sneaky" and don't
-/// report back to the C++ layer CQ operations
-class CallOpSetCollectionInterface
-    : public std::enable_shared_from_this<CallOpSetCollectionInterface> {};
+/// of the group should reference the collection, as will the object
+/// that instantiates the collection, allowing for ref-counting.
+/// Any actual use should derive from this base class. This is specifically
+/// necessary if some of the CallOpSet's in the collection are "Sneaky" and
+/// don't report back to the C++ layer CQ operations
+class CallOpSetCollectionInterface {
+ public:
+  CallOpSetCollectionInterface() {
+    gpr_atm_rel_store(&refs_, static_cast<gpr_atm>(1));
+  }
+  // always allocated against a call arena, no memory free required
+  static void operator delete(void* ptr, std::size_t size) {
```

Comment:
So, interestingly hiding the delete doesn't seem to work because of the magic of ["deleting destructors"](https://www.google.com/search?q=deleting+destructor) . There may be some compilers that don't use these as they're not required by the C++ spec, but GCC and Clang do. That said, yeah, I appreciate that we can't assert size in the delete of a non-final class, I think.